lib/hypothesis.rb

```diff
@@ -0,0 +1,108 @@
+module Hypothesis
+  class Error < Exception
+  end
+
+  class UnsatisfiedAssumption < Error
+  end
+
+  class UsageError < Error
+  end
+
+  def hypothesis(options={}, &block)
+    if defined? @current_hypothesis_engine and not @current_hypothesis_engine.nil?
+      raise UsageError.new("Cannot nest hypothesis calls")
+    end
+    begin
+      @current_hypothesis_engine = Engine.new(options) 
+      @current_hypothesis_engine.run(&block)
+    ensure
+      @current_hypothesis_engine = nil
+    end
+  end
+
+  def given(provider=nil, &block)
+    if not defined? @current_hypothesis_engine or @current_hypothesis_engine.nil?
+      raise UsageError.new("Cannot call given outside of a hypothesis block")
+    end
+    @current_hypothesis_engine.current_source.given(provider, &block)
+  end
+
+  def assume(condition)
+    if not defined? @current_hypothesis_engine or @current_hypothesis_engine.nil?
+      raise UsageError.new("Cannot call assume outside of a hypothesis block")
+    end
+    @current_hypothesis_engine.current_source.assume(condition)
+  end 
+
+  class Engine
+    attr_reader :current_source
+
+    def initialize(options)
+      @max_examples = options.delete(:max_examples) || 200
```

Comment:
What do you think about doing `options.fetch(:max_examples, 200)`  here?

Reply:
I think that it's been a while since I've written Ruby and I'd forgotten what that method was called, thanks. :-)